backend/protocol_rpc/message_handler/base.py

```diff
@@ -1,65 +1,66 @@
 
 
 class MessageHandler:
-
-    status_mappings = {
-        "debug": "info",
-        "info": "info",
-        "success": "info",
-        "error": "error",
-    }
-
-    def __init__(self, app: Flask, socketio: SocketIO, config: GlobalConfiguration):
-        self.app = app
+    def __init__(self, socketio: SocketIO, config: GlobalConfiguration):
         self.socketio = socketio
         self.config = config
         setup_logging_config()
 
     def log_endpoint_info(self, func):
         return log_endpoint_info_wrapper(self, self.config)(func)
 
-    def socket_emit(self, message):
-        self.socketio.emit("status_update", {"message": message})
-
-    def log_message(self, function_name, result: EndpointResult):
-        logging_status = self.status_mappings[result.status.value]
-        if hasattr(self.app.logger, logging_status):
-            log_method = getattr(self.app.logger, logging_status)
-            result_string = str(result)
-            function_result = (
-                (result_string[:MAX_LOG_MESSAGE_LENGTH] + "...")
-                if result_string is not None
-                and len(result_string) > MAX_LOG_MESSAGE_LENGTH
-                else result_string
-            )
-            log_method(function_name + ": " + function_result)
-        else:
-            raise Exception(f"Logger does not have the method {result.status}")
+    def socket_emit(self, log_event: LogEvent):
+        self.socketio.emit(log_event.name, log_event.to_dict())
+
+    def log_message(self, log_event: LogEvent):
+        logging_status = log_event.type.value
+
+        if not hasattr(logger, logging_status):
+            logging_status = "info"
+
+        log_method = getattr(logger, logging_status)
 
-    def send_message(
-        self,
-        function_name: str,
-        result: EndpointResult,
-    ):
-        self.log_message(function_name, result)
+        message = (
+            (log_event.message[:MAX_LOG_MESSAGE_LENGTH] + "...")
+            if log_event.message is not None
+            and len(log_event.message) > MAX_LOG_MESSAGE_LENGTH
+            else log_event.message
+        )
 
-        formatted_response = format_response(function_name, result)
-        self.socket_emit(formatted_response.to_json())
+        log_message = f"[{log_event.scope.value}] {message}"
+        gray = "\033[38;5;245m"
+        reset = "\033[0m"
+
+        if log_event.data:
+            try:
+                data_str = json.dumps(log_event.data, default=lambda o: o.__dict__)
+                log_message += f" {gray}{data_str}{reset}"
+            except TypeError as e:
+                log_message += (
+                    f" {gray}{str(log_event.data)} (serialization error: {e}){reset}"
+                )
+
+        log_method(log_message)
+
+    def send_message(self, log_event: LogEvent, log_to_terminal: bool = True):
+        if log_to_terminal:
+            self.log_message(log_event)
+        self.socket_emit(log_event)
```

Comment:
nit: maybe we can use loguru sinks to handle multiple outputs https://loguru.readthedocs.io/en/stable/api/logger.html#sink

Reply:
I though about it but unfortunately a sink only takes a Message, meaning I would lose the structuring of the data (LogEvent) and thus have to re-extract the data from the message somehow.